internal/rpc/httpproxy.go

```diff
@@ -7,98 +7,117 @@ import (
 	"crypto/tls"
 	"crypto/x509"
 	"fmt"
-	"io"
+	"math/rand"
 	"net/http"
+	"net/http/httputil"
 	"net/url"
+	"strings"
+	"time"
 
-	"github.com/juju/zaputil"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/names/v4"
 	"github.com/juju/zaputil/zapctx"
-	"gopkg.in/errgo.v1"
+	"go.uber.org/zap"
 
 	"github.com/canonical/jimm/v3/internal/dbmodel"
 )
 
-type httpOptions struct {
-	TLSConfig *tls.Config
-	URL       url.URL
+const (
+	defaultScheme = "https"
+)
+
+type ControllerProxy struct {
+	Controller dbmodel.Controller
+	Username   string
+	Password   string
 }
 
-// ProxyHTTP proxies the request to the controller using the info contained in dbmodel.Controller.
-// It tries for a controller, if it errors, it logs the error and go to the next, if no controller responds it returns a 504.
-func ProxyHTTP(ctx context.Context, ctl *dbmodel.Controller, w http.ResponseWriter, req *http.Request) error {
+func ProxyHTTP(ctx context.Context, ctl ControllerProxy, w http.ResponseWriter, req *http.Request) {
+	urls, err := getControllerAddresses(ctl.Controller)
+	if err != nil {
+		zapctx.Error(ctx, "failed to get controller addresses", zap.Error(err))
+		http.Error(w, fmt.Sprintf("failed to get controller addresses: %v", err), http.StatusInternalServerError)
+		return
+	}
+
 	var tlsConfig *tls.Config
-	if ctl.CACertificate != "" {
+	if ctl.Controller.CACertificate != "" {
 		cp := x509.NewCertPool()
-		ok := cp.AppendCertsFromPEM([]byte(ctl.CACertificate))
+		ok := cp.AppendCertsFromPEM([]byte(ctl.Controller.CACertificate))
 		if !ok {
 			zapctx.Warn(ctx, "no CA certificates added")
 		}
 		tlsConfig = &tls.Config{
 			RootCAs:    cp,
-			ServerName: ctl.TLSHostname,
+			ServerName: ctl.Controller.TLSHostname,
 			MinVersion: tls.VersionTLS12,
 		}
 	}
 
-	if ctl.PublicAddress != "" {
-		err := doRequest(ctx, w, req, httpOptions{
-			TLSConfig: tlsConfig,
-			URL:       createURLWithNewHost(*req.URL, ctl.PublicAddress),
-		})
-		if err == nil {
-			return nil
-		}
+	// transport is the default HTTP transport config with TLS configuration.
+	transport := &http.Transport{
+		TLSClientConfig:       tlsConfig,
+		Proxy:                 http.ProxyFromEnvironment,
+		ForceAttemptHTTP2:     true,
+		MaxIdleConns:          100,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   10 * time.Second,
+		ExpectContinueTimeout: 1 * time.Second,
 	}
```

Comment:
Can change this to:
```go
	transport := http.DefaultTransport.(*http.Transport).Clone()
	transport.TLSClientConfig = tlsConfig
```

Reply:
But we're using the default anyway and there's built in code to clone the default transport? I disagree, let's use the cloning built into the transport and just update the config rather than setting it all manually.